pennylane/ops/op_math/evolve.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-few-public-methods,function-redefined
+
+"""
+This file contains the ``ParametrizedEvolution`` operator.
+"""
+
+from typing import Union
+
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation, Operator
+
+from .exp import Evolution
+from .parametrized_hamiltonian import ParametrizedHamiltonian
+
+has_jax = True
+try:
+    import jax.numpy as jnp
+    from jax.experimental.ode import odeint
+except ImportError as e:
+    has_jax = False
+
+
+def evolve(op: Union[Operator, ParametrizedHamiltonian]):
+    """Returns a new operator that computes the evolution of ``op``.
+
+    Args:
+        op (Union[.Operator, .ParametrizedHamiltonian]): operator to evolve
+
+    Returns:
+        Union[.Evolution, ~pennylane.ops.op_math.evolve.ParametrizedEvolution]: evolution operator
+
+    .. seealso:: :class:`.ParametrizedEvolution`
+    .. seealso:: :class:`.Evolution`
+
+    **Examples**
+
+    We can use ``qml.evolve`` to compute the evolution of any PennyLane operator:
+
+    >>> op = qml.s_prod(2, qml.PauliX(0))
+    >>> qml.evolve(op)
+    Exp((-0-1j) 2*(PauliX(wires=[0])))
```

Comment:
```suggestion
    Exp(1j 2*(PauliX(wires=[0])))
```

Reply:
The reason that this is what it returns is that we expect users to absorb the sign into the coefficient or generator. I'm not sure where exactly, but my understanding is there was some discussion of what the convention should be in PennyLane, and it was decided that the convention is `exp(ixG)`.

Reply:
As a related question, is it strange to be able to create an `Evolution` operator in this way, but not have the option to define the coefficient? Currently I can't take some operator `H` and create the time evolution `exp(-iHt)` with `qml.evolve`. 

I think I would expect this to either be specific to `ParametrizedHamiltonian`, or to create an `Evolution` operator in the same way that other functions in `math_ops` that construct a `SymbolicOp` work, i.e. that it can also has arguments `param`, `do_queue=True`, `id=None` and uses those in the case where it takes and `Operator` and returns a `SymbolicOp`.

Reply:
I opened a new PR fixing the docstring: https://github.com/PennyLaneAI/pennylane/pull/3656

To create

Reply:
The problem is that we want `qml.evolve` to be as generic as possible (we want to return `Evolution` and `ParametrizedEvolution` operators). 

We can add the extra kwargs for both operators, but I think it might get confusing for users, given that in some cases those kwargs will be ignored (e.g. `t` kwarg when NOT using a `ParametrizedHamiltonian`)

Reply:
I 100% agree about it being confusing if we add the extra kwargs, but I guess I'm wondering, if we don't think it makes sense to do that, whether we should just accept that these are two substantially different things with, ideally, substantially different arguments, and they shouldn't be combined into one function. 

Wrt your example, I can see that you create an Evolution operator in this way, but a primary reason for having the `Evolution` operator instead of just using `Exp` is to have the trainable parameter. Here, you initialize an `Evolution` operator with a parameter of 1, no matter what. I.e. technically you created an operator of the form `exp(-iHt)`, and just set `t=1` in your initial initialization, but the trainable parameter seems quite unclear/hidden. This is made more confusing by the fact that most of the other functions to create SymbolicOps all work the same way, while this one is different. 

At the very least we should add much clearer documentation about how to create an Evolution operator like this, either here or on Evolution (or a bit of both, probably). But personally I'm not convinced the generalization is worth the lack of clarity here.

Reply:
Pinging @trbromley here

Reply:
Just thinking out loud: do we really need the `coeff`/`param` attributes in `Exp`/`Evolution`?

Wouldn't it be simpler if `Exp` only exponentiates an operator? Why adding a scalar product? A user can already do this before exponentiating the operator:
```pycon
>>> qml.exp(qml.s_prod(4, qml.PauliX(0)))  # hopefully we can do 4 * qml.PauliX(0) soon
Exp(4*(PauliX(wires=[0])))
```

Reply:
@albi3ro any thoughts? 🤔